crdexample/pkg/resource.go

```diff
@@ -0,0 +1,138 @@
+
+var log = logutil.Log.WithName("firstmate-resource")
+
+// firstMate is the Schema for the firstmates API
+// +k8s:openapi-gen=true
+type FirstMate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   FirstMateSpec   `json:"spec,omitempty"`
+	Status FirstMateStatus `json:"status,omitempty"`
+}
+
+// FirstMateSpec defines the desired state of firstMate
+type FirstMateSpec struct {
+	Height     *int  `json:"height,omitempty"`
+	Experience int   `json:"experience,omitempty"`
+	Crew       int32 `json:"crew,omitempty"`
+}
+
+// FirstMateStatus defines the observed state of firstMate
+type FirstMateStatus struct {
+	Location string
+}
+
+var _ webhook.Validator = &FirstMate{}
+
+// ValidateCreate implements webhookutil.validator so a webhook will be registered for the type
+func (f *FirstMate) ValidateCreate() error {
+	log.Info("validate create", "name", f.Name)
+
+	if f.Spec.Crew <= 0 {
+		return fmt.Errorf("crew must be greater than 0")
+	}
+	return nil
+}
+
+// ValidateUpdate implements webhookutil.validator so a webhook will be registered for the type
+func (f *FirstMate) ValidateUpdate(old runtime.Object) error {
+	log.Info("validate update", "name", f.Name)
+
+	if f.Spec.Crew <= 0 {
+		return fmt.Errorf("crew must be greater than 0")
+	}
+
+	oldF, ok := old.(*FirstMate)
+	if !ok {
+		return fmt.Errorf("expect old object to be a %T instead of %T", oldF, old)
+	}
+	if f.Spec.Crew-oldF.Spec.Crew > 1 || oldF.Spec.Crew-f.Spec.Crew > 1 {
+		return fmt.Errorf("crew must be greater than 0")
+	}
+	return nil
+}
+
+var _ webhook.Defaulter = &FirstMate{}
+
+// Default implements webhookutil.defaulter so a webhook will be registered for the type
+func (f *FirstMate) Default() {
+	log.Info("default", "name", f.Name)
+
+	if *f.Spec.Height == 0 {
+		height := 10
+		f.Spec.Height = &height
+	}
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FirstMateList contains a list of firstMate
+type FirstMateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []FirstMate `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&FirstMate{}, &FirstMateList{})
+}
+
+var (
+	// SchemeGroupVersion is group version used to register these objects
+	SchemeGroupVersion = schema.GroupVersion{Group: "crew.example.com", Version: "v1"}
+
+	// SchemeBuilder is used to add go types to the GroupVersionKind scheme
+	SchemeBuilder = &scheme.Builder{GroupVersion: SchemeGroupVersion}
+
+	// AddToScheme is required by pkg/client/...
+	AddToScheme = SchemeBuilder.AddToScheme
+)
+
+// Resource is required by pkg/client/listers/...
+func Resource(resource string) schema.GroupResource {
+	return SchemeGroupVersion.WithResource(resource).GroupResource()
+}
+
+// TODO: figure out should we encourage the users to do a wrapper for a core type object.
+type MyPod struct {
+	corev1.Pod
+}
```

Comment:
Its not clear to me in this case whether the webhook gets generated for the `corev1.Pods` (the embedded type) or the wrapped type ?

Reply:
This was some experimental code, but they actually sneak in.
I will remove this piece.
Sorry for the confusion.